Deb does not build anymore in Ubuntu 24.04 because of a memcopy problem

The deb package does not build anymore on ubuntu 24.04 because a problem in memcopy [openzim/libzim/actions/runs/14224960852/job/39862160011?pr=958#step:12:459](https://github.com/openzim/libzim/actions/runs/14224960852/job/39862160011?pr=958#step:12:459)

@kelson42 
I spun up an ubuntu 24.04 container and matched the build options to try to replicate the issue, and I was unable to. The main thing of note between ubuntu 24.04 and prior ubuntu version is 24.04 added `'-D_FORTIFY_SOURCE=3'`. The ubuntu 24.04 container may just need to be rebuilt, as the gcc/glibc bug trackers had some hits for `D_FORTIFY_SOURCE=3` giving false positives that were subsequently patched. As a side note, I did mention this issue with you 2025-Feb-20 on slack, but I should have followed up. My bad. 

```meson
project('libzim', ['c', 'cpp'],
  version : '9.2.3',
  license : 'GPL2',
  default_options : ['c_std=c11', 'cpp_std=c++17', 'werror=true', 'b_lto=true'])

add_project_arguments(
  '-O2',
  '-g',
  '-fno-omit-frame-pointer',
  '-mno-omit-leaf-frame-pointer',
  '-fstack-protector-strong',
  '-fstack-clash-protection',
  '-Wformat',
  '-Werror=format-security',
  '-fcf-protection',
  '-Wdate-time',
  '-D_FORTIFY_SOURCE=3',
  '-Wall',
  '-Winvalid-pch',
  '-D_LARGEFILE64_SOURCE=1',
  '-D_FILE_OFFSET_BITS=64',
  language: 'cpp')

```

For completeness, this is the error that is thrown in the deb build:
```
c++ -Itest/tinyString.p -Iinclude -I../include -Isrc -I../src -fdiagnostics-color=always -Wall -Winvalid-pch -Werror -std=c++17 -D_LARGEFILE64_SOURCE=1 -D_FILE_OFFSET_BITS=64 -DLIBZIM_EXPORT_PRIVATE_DLL -g -O2 -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -ffile-prefix-map=/github/workspace=. -flto=auto -ffat-lto-objects -fstack-protector-strong -fstack-clash-protection -Wformat -Werror=format-security -fcf-protection '-fdebug-prefix-map=/github/workspace=/usr/src/libzim-9.2.3+git202502152242.ea8b057~noble' -Wdate-time -D_FORTIFY_SOURCE=3 -DGTEST_HAS_PTHREAD=1 -pthread -DWITH_TEST_DATA=0 -MD -MQ test/tinyString.p/tinyString.cpp.o -MF test/tinyString.p/tinyString.cpp.o.d -o test/tinyString.p/tinyString.cpp.o -c ../test/tinyString.cpp
In file included from ../test/../src/writer/tinyString.h:23,
                 from ../test/tinyString.cpp:22:
In function ‘zim::safe_memcmp<char>(char const*, char const*, unsigned long)int’,
    inlined from ‘zim::writer::TinyString::operator<(zim::writer::TinyString const&) const’ at ../test/../src/writer/tinyString.h:74:39,
    inlined from ‘(anonymous namespace)::TinyStringTest_chars_Test::TestBody()’ at ../test/tinyString.cpp:65:3:
../test/../src/writer/../tools.h:84:25: error: ‘memcmp’ specified bound [1, 65534] exceeds source size 0 [-Werror=stringop-overread]
   84 |       return std::memcmp(lhs, rhs, count);
      |              ~~~~~~~~~~~^~~~~~~~~~~~~~~~~
In function ‘zim::safe_memcmp<char>(char const*, char const*, unsigned long)int’,
    inlined from ‘zim::writer::TinyString::operator<(zim::writer::TinyString const&) const’ at ../test/../src/writer/tinyString.h:74:39,
    inlined from ‘(anonymous namespace)::TinyStringTest_chars_Test::TestBody()’ at ../test/tinyString.cpp:70:3:
../test/../src/writer/../tools.h:84:25: error: ‘memcmp’ specified bound [1, 65534] exceeds source size 0 [-Werror=stringop-overread]
   84 |       return std::memcmp(lhs, rhs, count);
      |              ~~~~~~~~~~~^~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
```

--- a/src/tools.h
+++ b/src/tools.h
@@ -22,8 +22,6 @@
 #ifndef OPENZIM_LIBZIM_TOOLS_H
 #define OPENZIM_LIBZIM_TOOLS_H
 
-#include <cstddef> // size_t
-#include <cstring> // memcmp
 #include <string>
 #include <tuple>
 #include <map>
@@ -71,17 +69,6 @@ namespace zim {
       }
     }
     return count;
-  }
-
-  // Work around for errors reported by too picky UB sanitizer tools on std::memcmp()
-  // being called with a nullptr pointer argument even though count==0. Explicit
-  // nullptr checks to satisfy other picky compilers.
-  template <typename T>
-  int safe_memcmp(T const *const lhs, T const *const rhs, std::size_t count) noexcept {
-      if(count == 0UL || lhs == nullptr || rhs == nullptr) {
-          return 0;
-      }
-      return std::memcmp(lhs, rhs, count);
   }
 
   namespace writer {

--- a/src/writer/tinyString.h
+++ b/src/writer/tinyString.h
@@ -20,7 +20,6 @@
 #ifndef ZIM_WRITER_TINYSTRING_H
 #define ZIM_WRITER_TINYSTRING_H
 
-#include "../tools.h"
 #include "../zim_types.h"
 #include <cstring>
 
@@ -67,11 +66,11 @@ namespace zim
         size_t size() const { return m_size; }
         const char* const data() const { return m_data; }
         bool operator==(const TinyString& other) const {
-          return (m_size == other.m_size) && (safe_memcmp(m_data, other.m_data, m_size) == 0);
+          return (m_size == other.m_size) && (std::memcmp(m_data, other.m_data, m_size) == 0);
         }
         bool operator<(const TinyString& other) const {
-          const auto min_size = std::min(m_size, other.m_size);
-          const auto ret = safe_memcmp(m_data, other.m_data, min_size);
+          auto min_size = std::min(m_size, other.m_size);
+          auto ret = std::memcmp(m_data, other.m_data, min_size);
           if (ret == 0) {
             return m_size < other.m_size;
           } else {

